File: 程序员面试金典第6版/面试题 08.08. Permutation II LCCI.py
# 72%
class Solution:
    def permutation(self, S: str) -> List[str]:
        counter = collections.Counter(S)
        res = []
        def backtrack(cur):
            if not counter:
                res.append(cur)
                return
            for c in list(counter.keys()):
                counter[c] -= 1
                if counter[c] == 0:
                    del counter[c]
                backtrack(cur+c)
                counter[c] = counter.get(c, 0) + 1
        backtrack("")
        return res

File: 程序员面试金典第6版/面试题 08.08. Permutation II LCCI.cpp
class Solution {
public:
    vector<string> permutation(string S) {
        sort(S.begin(), S.end());
        vector<string> res;
        res.push_back(S);
        while(next_permutation(S.begin(), S.end())) {
            res.push_back(S);
        }
        return res;
    }
};